jdaviz/core/tools.py

```diff
@@ -354,28 +354,88 @@ def activate(self):
         self.viewer.add_event_callback(self.on_mouse_event,
                                        events=['dragmove', 'click'])
         for mark in self.viewer.figure.marks:
-            if np.any([x in mark.labels for x in ('vmin', 'vmax')]):
+            if np.any([x in mark.labels for x in ('vmin', 'vmax', 'stretch_knots')]):
                 mark.colors = ["#c75d2c"]
 
     def deactivate(self):
         self.viewer.remove_event_callback(self.on_mouse_event)
         for mark in self.viewer.figure.marks:
-            if np.any([x in mark.labels for x in ('vmin', 'vmax')]):
+            if np.any([x in mark.labels for x in ('vmin', 'vmax', 'stretch_knots')]):
                 mark.colors = ["#007BA1"]
 
     def on_mouse_event(self, data):
         if (time.time() - self._time_last) <= 0.05:
             # throttle to 200ms
             return
 
+        # Mouse event coordinates
         event_x = data['domain']['x']
-        current_bounds = [self.viewer._plugin.stretch_vmin_value,
-                          self.viewer._plugin.stretch_vmax_value,]
-        att_names = ["stretch_vmin_value", "stretch_vmax_value"]
-        closest_bound_ind = np.argmin([abs(current_bounds[0] - event_x),
-                                       abs(current_bounds[1] - event_x)])
+        event_y = data['domain']['y']
+
+        x_min, x_max = self.viewer.state.x_min, self.viewer.state.x_max
+        y_min, y_max = self.viewer.state.y_min, self.viewer.state.y_max
 
-        setattr(self.viewer._plugin, att_names[closest_bound_ind], event_x)
+        current_bounds = [self.viewer._plugin.stretch_vmin_value,
+                          self.viewer._plugin.stretch_vmax_value]
+
+        distances_to_bounds = [abs(current_bounds[0] - event_x), abs(current_bounds[1] - event_x)]
+        closest_bound_index = np.argmin(distances_to_bounds)
+        closest_bound_distance = distances_to_bounds[closest_bound_index]
+
+        stretch_type = self.viewer._plugin.stretch_function_value
+        if stretch_type == 'spline':
+            # Retrieve current knots and bounds
+            # knot_x is in units of the x-axis (pixel value)
+            knot_x = self.viewer._plot.marks['stretch_knots'].x
+            # knot_y is in units of the (hidden) normalized y-axis from 0 to 1
+            knot_y = self.viewer._plot.marks['stretch_knots'].y
+
+            # event_y is in units of the y-axis (density),
+            # so we need to rescale event_y to the same units as knot_y
+            event_y_normalized = (event_y - y_min) / (y_max - y_min)
+            # Drop event if the event coordinate are out of bounds
+            if not 0 <= event_y_normalized <= 0.92:
+                return
+
+            # Distance from mouse position to each knot, normalized to viewer axes size
+            distances_to_knots = np.sqrt(((knot_x - event_x) / (x_max - x_min)) ** 2 +
+                                         ((knot_y - event_y_normalized) / (y_max - y_min)) ** 2)
```

Comment:
(btw I said removing the normalization seems correct because it does look like `knot_y` is already normalized, so `knot_y - y_normalized` is something like `0.1611574 - 0.22782`)

Reply:
Yes, the knot_y values are normalized. Great catch! Thanks for the sharp insights.

Reply:
Thank you both for your valuable insights. Ricky, I agree that throttling can help reduce the latency issue of updates in the knot up to a certain level. However, I understand that the computation doesn't work for edge cases like the first knot/bound that you described. Therefore, I will update the code to compare the x-distance for bounds and knots in such cases, instead of using radial vs x comparison. 
<img width="359" alt="image" src="https://github.com/spacetelescope/jdaviz/assets/66814693/d2bef901-339d-4648-80eb-d8834de28e43">